base/invalidation.jl

```diff
@@ -0,0 +1,164 @@
+        end
+        found_arg = false
+        ngrs = 0
+        for ur in Compiler.userefs(stmt)
+            arg = ur[]
+            # If any of the GlobalRefs in this stmt match the one that
+            # we are about, we need to move out all GlobalRefs to preserve
+            # effect order, in case we later invalidate a different GR
+            if isa(arg, GlobalRef)
+                ngrs += 1
+                if isgr(arg)
+                    @assert !isa(stmt, PhiNode)
+                    found_arg = found_any = true
+                    break
+                end
+            end
+        end
+        if found_arg
+            get_labelchangemap()[i] += ngrs
+        end
+    end
+    next_empty_idx = 1
+    if labelchangemap !== nothing
+        Compiler.cumsum_ssamap!(labelchangemap)
+        new_stmts = Vector(undef, length(stmts)+labelchangemap[end])
+        new_ssaflags = Vector{UInt32}(undef, length(new_stmts))
+        new_debuginfo = Compiler.DebugInfoStream(nothing, src.debuginfo, length(new_stmts))
+        new_debuginfo.def = src.debuginfo.def
+        for i = 1:length(stmts)
+            stmt = stmts[i]
+            urs = Compiler.userefs(stmt)
+            new_stmt_idx = i+labelchangemap[i]
+            for ur in urs
+                arg = ur[]
+                if isa(arg, SSAValue)
+                    ur[] = SSAValue(arg.id + labelchangemap[arg.id])
+                elseif next_empty_idx != new_stmt_idx && isa(arg, GlobalRef)
+                    new_debuginfo.codelocs[3next_empty_idx - 2] = i
+                    new_stmts[next_empty_idx] = arg
+                    new_ssaflags[next_empty_idx] = UInt32(0)
+                    ur[] = SSAValue(next_empty_idx)
+                    next_empty_idx += 1
+                end
+            end
+            @assert new_stmt_idx == next_empty_idx
+            new_stmts[new_stmt_idx] = urs[]
+            new_debuginfo.codelocs[3new_stmt_idx - 2] = i
+            new_ssaflags[new_stmt_idx] = src.ssaflags[i]
+            next_empty_idx = new_stmt_idx+1
+        end
+        src.code = new_stmts
+        src.ssavaluetypes = length(new_stmts)
+        src.ssaflags = new_ssaflags
+        src.debuginfo = Core.DebugInfo(new_debuginfo, length(new_stmts))
+    end
+    return found_any
+end
+
+function invalidate_code_for_globalref!(gr::GlobalRef, new_max_world::UInt)
+    valid_in_valuepos = false
+    foreach_reachable_mtable(new_max_world) do mt::Core.MethodTable
+        for method in MethodList(mt)
+            if isdefined(method, :source)
+                src = _uncompressed_ir(method)
+                old_stmts = src.code
+                if invalidate_code_for_globalref!(gr, src)
+                    if src.code !== old_stmts
+                        method.debuginfo = src.debuginfo
+                        method.source = src
+                        method.source = ccall(:jl_compress_ir, Ref{String}, (Any, Ptr{Cvoid}), method, C_NULL)
+                    end
```

Comment:
I don't love that we are adding a lot of data races here, and still not even getting the result correct, since lowering happens a long time before taking world_counter_lock (commonly in an entirely different process), so anything that got loaded from a package image afterwards (which is most things) will not be able to be handled correctly. Could we fix lowering/inference not to need these attempted and unreliable hacks?

Reply:
#56714